### PR TITLE
`exp push/pull`: add `--rev`, `-n`, `-A` flags

### PR DESCRIPTION
**Pull request by karajan1001**
Related to https://github.com/iterative/dvc/pull/7255

> & https://github.com/iterative/dvc/pull/7245

### CODE REVIEW

**Comment by jorgeorpinel** [comment]
Is it possible to re-do this PR from a branch directly on this repo? Thanks

File: content/docs/command-reference/exp/pull.md [hunk 1]
```diff
@@ -43,6 +43,14 @@ all <abbr>cached</abbr> data associated with the experiment to DVC
 
 ## Options
 
+- `-A`, `--all-commits` - Pull all experiments in the repository (overrides
+  `--rev`).
+
+- `--rev <commit>` - Pull experiments derived from the specified Git commit.
+
+- `-n <num>`, `--num <num>` - Pull experiments from the `<rev>` commit and `num`
+  commits before it. (`-1` for the whole history of the `<rev>`)
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/pull.md:52]: Same as https://github.com/iterative/dvc.org/pull/3187#pullrequestreview-864359473

File: content/docs/command-reference/exp/pull.md [hunk 2]
```diff
@@ -43,6 +43,14 @@ all <abbr>cached</abbr> data associated with the experiment to DVC
 
 ## Options
 
+- `-A`, `--all-commits` - Pull all experiments in the repository (overrides
+  `--rev`).
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/pull.md:47]: Per https://github.com/iterative/dvc/pull/7245

```suggestion
- `-A`, `--all-commits` - Pull all experiments in the repository (overrides
  `--rev` and `--num`).
```

**Comment by jorgeorpinel** [comment]
> Is it possible to re-do this PR from a branch directly on this repo? Thanks

UPDATE: Never mind, I reviewed this already.

File: content/docs/command-reference/exp/pull.md [hunk 3]
```diff
@@ -43,6 +43,15 @@ all <abbr>cached</abbr> data associated with the experiment to DVC
...
 ## Options
 
+- `-A`, `--all-commits` - pull all experiments in the repository (overrides
+  `--rev` and `--num`).
+
+- `--rev <commit>` - pull experiments derived from the specified `<commit>` as
+  baseline.
+
+- `-n <num>`, `--num <num>` - pull experiments from the `--rev` baseline and
+  from `num` commits before it. Give `-1` to include all the parent commits.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/pull.md:53]: Here

File: content/docs/command-reference/exp/push.md [hunk 4]
```diff
@@ -40,6 +40,15 @@ This command will also try to [push](/doc/command-reference/push) all
...
 ## Options
 
+- `-A`, `--all-commits` - push all experiments in the repository (overrides
+  `--rev` and `--num`).
+
+- `--rev <commit>` - push experiments derived from the specified `<commit>` as
+  baseline.
+
+- `-n <num>`, `--num <num>` - push experiments from the `--rev` baseline and
+  from `num` commits before it. Give `-1` to include all the parent commits.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/push.md:50]: and here

**Review by jorgeorpinel — COMMENTED** [review]
These will need an update again once https://github.com/iterative/dvc.org/pull/3187#pullrequestreview-883803230 is resolved 🙂

File: content/docs/command-reference/exp/pull.md [hunk 5]
```diff
@@ -43,6 +43,16 @@ all <abbr>cached</abbr> data associated with the experiment to DVC
...
 
+- `-A`, `--all-commits` - pull all experiments in the repository (overrides
+  `--rev` and `--num`).
+
+- `--rev <commit>` - pull experiments derived from the specified `<commit>` as
+  baseline.
+
+- `-n <num>`, `--num <num>` - show experiments from the previous `num` commits
+  (before the `--rev` baseline). For example `-n -2` includes experiments from
+  `HEAD` and its 2 first ancestors.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/pull.md:54]: See https://github.com/iterative/dvc.org/pull/3187#pullrequestreview-902553693

File: content/docs/command-reference/exp/push.md [hunk 6]
```diff
@@ -40,6 +40,16 @@ This command will also try to [push](/doc/command-reference/push) all
...
 
+- `-A`, `--all-commits` - push all experiments in the repository (overrides
+  `--rev` and `--num`).
+
+- `--rev <commit>` - push experiments derived from the specified `<commit>` as
+  baseline.
+
+- `-n <num>`, `--num <num>` - show experiments from the previous `num` commits
+  (before the `--rev` baseline). For example `-n -2` includes experiments from
+  `HEAD` and its 2 first ancestors.
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/push.md:51]: Same

File: content/docs/command-reference/exp/push.md [hunk 7]
```diff
@@ -40,6 +40,16 @@ This command will also try to [push](/doc/command-reference/push) all
...
 
+- `-A`, `--all-commits` - push all experiments in the repository (overrides
+  `--rev` and `--num`).
+
+- `--rev <commit>` - push experiments derived from the specified `<commit>` as
+  baseline.
+
+- `-n <num>`, `--num <num>` - show experiments from the `--rev` baseline and
+  from `num` commits before it (first parents). Give a negative value to include
+  all first-parent commits (similar to `git log -n`).
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/push.md:51]: Same as https://github.com/iterative/dvc.org/pull/3187#pullrequestreview-903319359, https://github.com/iterative/dvc.org/pull/3191#pullrequestreview-903953082
```suggestion
- `-n <num>`, `--num <num>` - show experiments from the last `num` commits
  (first parents) starting from the `--rev` baseline. Give a negative value to include
  all first-parent commits (similar to `git log -n`).
```

File: content/docs/command-reference/exp/pull.md [hunk 8]
```diff
@@ -43,6 +43,16 @@ all <abbr>cached</abbr> data associated with the experiment to DVC
...
 
+- `-A`, `--all-commits` - pull all experiments in the repository (overrides
+  `--rev` and `--num`).
+
+- `--rev <commit>` - pull experiments derived from the specified `<commit>` as
+  baseline.
+
+- `-n <num>`, `--num <num>` - show experiments from the `--rev` baseline and
+  from `num` commits before it (first parents). Give a negative value to include
+  all first-parent commits (similar to `git log -n`).
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/pull.md:54]: Same as https://github.com/iterative/dvc.org/pull/3187#pullrequestreview-903319359, https://github.com/iterative/dvc.org/pull/3191#pullrequestreview-903953082
```suggestion
- `-n <num>`, `--num <num>` - show experiments from the last `num` commits
  (first parents) starting from the `--rev` baseline. Give a negative value to include
  all first-parent commits (similar to `git log -n`).
```

File: content/docs/command-reference/exp/pull.md [hunk 9]
```diff
@@ -6,13 +6,13 @@ data from a `dvc remote`.
...
-usage: dvc exp pull [-h] [-q | -v] [-f] [--no-cache]
-                    [-r <name>] [-j <number>] [--run-cache]
+usage: dvc exp pull [-h] [-q | -v] [-A] [--rev <commit>] [-n <num>] [-f]
+                    [--no-cache] [-r <name>] [-j <number>] [--run-cache]
                     git_remote experiment
 
 positional arguments:
   git_remote          Git remote name or Git URL
-  experiment          Experiment to pull
+  experiment          Experiments to pull
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/pull.md:15]: If it's experiments shouldn't the usage block look different?

File: content/docs/command-reference/exp/push.md [hunk 10]
```diff
@@ -6,13 +6,13 @@ to a `dvc remote`.
...
-usage: dvc exp push [-h] [-q | -v] [-f] [--no-cache]
-                    [-r <name>] [-j <number>] [--run-cache]
+usage: dvc exp push [-h] [-q | -v] [-A] [--rev <commit>] [-n <num>] [-f]
+                    [--no-cache] [-r <name>] [-j <number>] [--run-cache]
                     git_remote experiment
 
 positional arguments:
   git_remote          Git remote name or Git URL
-  experiment          Experiment to push
+  experiment          Experiments to push
```
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/push.md:15]: Same Q

File: content/docs/command-reference/exp/pull.md (hunk 9 above)
Comment (dberenbaum) [reply content/docs/command-reference/exp/pull.md:15]: This is what I get for `dvc exp pull --help`:

```
usage: dvc experiments pull [-h] [-q | -v] [-A] [--rev <commit>] [-n <num>] [-f]
                            [--no-cache] [-r <name>] [-j <number>] [--run-cache]
                            <git_remote> [<experiment> ...]
```

Is there any reason we wouldn't use that? Are others not seeing the same?

Comment (jorgeorpinel) [reply content/docs/command-reference/exp/pull.md:15]: Yes what you see is what I meant @dberenbaum , specifically the line

`                            <git_remote> [<experiment> ...]`

But if this is part of #3340 and other refs need to be updated as well, maybe don't include the change here at all.

File: content/docs/command-reference/exp/push.md (hunk 10 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/push.md:15]: UPDATE: See https://github.com/iterative/dvc/pull/7206#pullrequestreview-924091543

Comment (karajan1001) [reply content/docs/command-reference/exp/push.md:15]: Also change line 11 to ` git_remote [experiment [experiment ...]]`?

File: content/docs/command-reference/exp/pull.md (hunk 9 above)
Comment (jorgeorpinel) [inline content/docs/command-reference/exp/pull.md:15]: Would prefer to discuss this separately. Seems out of scope for this already long-standing PR.
```suggestion
  experiment          Experiment to pull
```

File: content/docs/command-reference/exp/push.md (hunk 10 above)
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/push.md:15]: Idk, can we discuss that in a separate PR? Ideally in the core PR related to this. Thanks

Comment (jorgeorpinel) [inline content/docs/command-reference/exp/push.md:15]:
```suggestion
  experiment          Experiment to push
```

**Review by jorgeorpinel — CHANGES_REQUESTED** [review]
☝🏼 Thanks

**Review by jorgeorpinel — APPROVED** [review]
Merging!
